README.md
# Lengow for Magento

- **Requires at least:** 1.9
- **Tested up to:** 1.9
- **Requires PHP:** 5.4
- **Stable tag:** 3.4.1
- **License:** OSL-3.0
- **License URI:** https://opensource.org/licenses/OSL-3.0

## Overview

<p align="center">
  <img src="https://my.lengow.io/images/pages/launching/orders.png">
</p>

Lengow is the e-commerce automation solution that helps brands and distributors improve their performance, automate their business processes, and grow internationally. The Lengow platform is the key to strong profitability and visibility for products sold by online retailers around the world on all distribution channels: marketplaces, comparison shopping engines, affiliate platforms and display/retargeting platforms. Since 2009, Lengow has integrated more than 1,600 partners into its solution to provide a powerful platform to its 4,600 retailers and brands in 42 countries around the world.

Major features in Lengow include:

- Easily import your product data from your cms
- Use Lengow to target and exclude the right products for the right channels and tools (marketplaces, price comparison engines, product ads, retargeting, affiliation) and automate the process of product diffusion.
- Manipulate your feeds (categories, titles, descriptions, rules…) - no need for technical knowledge.
- Lengow takes care of the centralisation of orders received from marketplaces and synchronises inventory data with your backoffice. Track your demands accurately and set inventory rules to avoid running out of stock.
- Monitor and control your ecommerce activity using detailed, yet easy to understand graphs and statistics. Track clicks, sales, CTR, ROI and tweak your campaigns with automatic rules according to your cost of sales / profitability targets.
- Thanks to our API, Lengow is compatible with many applications so you can access the functionality of all your ecommerce tools on a single platform. There are already more than 40 available applications: marketing platform, translation, customer review, email, merchandise, price watch, web-to-store, product recommendation and many more

The Lengow plugin is free to download and it enables you to export your product catalogs and manage your orders. It is compatible only with the new version of our platform.
A Lengow account is created during the extension installation and you will have free access to our platform for 15 days. To benefit from all the functionalities of Lengow, this requires you to pay for an account on the Lengow platform.

## Plugin installation

Follow the instruction below if you want to install Lengow for Magento using Git.

1.) Clone the git repository to the desired location using:

    git clone git@github.com:lengow/plugin-magento.git Lengow_Export

In case you wish to contribute to the plugin, fork the `dev` branch rather than cloning it, and create a pull request via Github. For further information please read the section "Become a contributor" of this document.

2.) Set the correct directory permissions:

    chmod -R 755 Lengow_Export

Depending on your server configuration, it might be necessary to set whole write permissions (777) to the files and folders above.
You can also start testing with lower permissions due to security reasons (644 for example) as long as your php process can write to those files.

3.) Run the install script in `tools` folder to creates symbolic links to the sources of the plugin:
    
    sh install.sh magento_install_path
    
`magento_install_path` represents the path to the root folder of your Magento installation (folder containing the app, skin, var ... folders).

4.) Activate the option 'symlinks' in the Magento configuration (Configuration > Advanced > Developer > Template Settings Allow Simlinks => Yes).

Warning ! On the latest 1.9 versions of Magento, this option was removed for security reasons. Connect to your database and play the following SQL query:

    INSERT INTO core_config_data (config_id, scope, scope_id, path, value) VALUES (NULL , 'default', '0', 'dev/template/allow_symlink', '1');

5.) Go to "System" > "Cache Management" and click both the "Flush Magento Cache" as well as the "Flush Cache Storage" button.

6.) Log out, then log back in on the Magento back-office.

4.) Log in with your Lengow credentials and configure the plugin.

## Frequently Asked Questions

### Where can I find Lengow documentation and user guides?

For help setting up and configuring Lengow plugin please refer to our [user guide](https://help.lengow.com/hc/en-us/articles/360011978332)

### Where can I get support?

To make a support request to Lengow, use [our helpdesk](https://help.lengow.com/hc/en-us/requests/new).


## Become a contributor

Lengow for Magento is available under license (OSL-3.0). If you want to contribute code (features or bugfixes), you have to create a pull request via Github and include valid license information.

The `master` branch contains the latest stable version of the plugin. The `dev` branch contains the version under development.
All Pull requests must be made on the `dev` branch and must be validated by reviewers working at Lengow.

By default the plugin is made to work on our pre-production environment (my.lengow.net).
To change this environment, you must modify the two constants present in the file `app/code/communtity/Lengow/Connector/Model/Connector.php`

    const LENGOW_URL = 'lengow.net';
    const LENGOW_API_URL = 'https://api.lengow.net';

### Translation

Translations in the plugin are managed via a key system and associated yaml files

Start by installing Yaml Parser:

    sudo apt-get install php5-dev libyaml-dev
    sudo pecl install yaml
    
To translate the project, use specific key in php code and modify the *.yml files in the directory: `app/code/communtity/Lengow/Connector/locale/yml/`

Once the translations are finished, just run the translation update script in `tools` folder

    php translate.php
    
The plugin is translated into English, French and German.

## Changelog

The changelog and all available commits are located under [CHANGELOG](CHANGELOG).


CHANGELOG
=============================================================
Version 3.4.2
=============================================================



=============================================================
Version 3.4.1
=============================================================

    - Feature: Adding the PHP version in the toolbox
    - Feature: Modification of the fallback urls of the Lengow Help Center
    - Feature: Adding extra field update date in external toolbox
    - BugFix: [Import] Loading of order types at each order synchronization

=============================================================
Version 3.4.0
=============================================================

    - Feature: Integration of order synchronization in the toolbox webservice
    - Feature: Retrieving the status of an order in the toolbox webservice

=============================================================
Version 3.3.1
=============================================================

    - Feature: Outsourcing of the toolbox via webservice
    - Feature: Setting up a modal for the plugin update

=============================================================
Version 3.3.0
=============================================================

    - Feature: Integration of the new connection process

=============================================================
Version 3.2.3
=============================================================

    - Feature: Adding new links to the Lengow Help Center and Support
    - Feature: B2B orders can now be imported without taxes (optional)
    - Bugfix: [import] Adding timezone in the order recovery date
    - BugFix: [import] Handling of b2b order shipping tax
    - BugFix: [import] Add missing param to 'toInvoice()' method call
    - BugFix: Correction of Customer::getName behavior
    - BugFix: [import] Changing variable name into checkAndChangeMarketplaceName() function
    - BugFix: Always load iframe over HTTPS

=============================================================
Version 3.2.2
=============================================================

    - Feature: [import] Addition of order types in the order management screen
    - Feature: [import] Integration of the region code in the delivery and billing addresses
    - Feature: [export] Add option to select which field should be taken from parent products
    - Bugfix: [import] Refactoring of the creation of delivery and billing addresses
	- Bugfix: [export] Fix an if statement in getShippingPrice to check all posibilities
	- Bugfix: Update of the access token when recovering an http 401 code

=============================================================
Version 3.2.1
=============================================================

    - Bugfix: Addition of the http 201 code in the success codes

=============================================================
Version 3.2.0
=============================================================

    - Feature: Refactoring and optimization of the connector class
    - Feature: [import] Protection of the import of anonymized orders
    - Feature: [import] Protection of the import for orders older than 3 months
    - Feature: Optimization of API calls for synchronisation of orders and actions
    - Feature: Display of an alert when the plugin is no longer up to date
    - Feature: Renaming from Preprod Mode to Debug Mode
    - Bugfix: Refactoring and optimization of dates with the correct locale
    - Bugfix: [action] Improved carrier matching with a strict then approximate search
    - Bugfix: [import] Enhanced security for orders that change their marketplace name

=============================================================
Version 3.1.3
=============================================================

    - Feature: Adding compatibility with php 7.3
    - Bugfix: [export] Improvement of the counter of the number of product to be exported
    - Bugfix: Catching of the error on the creation of the marketplaces.json file

=============================================================
Version 3.1.2
=============================================================

    - Feature: [action] Improved carrier matching with search on carrier code and label
    - Bugfix: [import] Save tracking number during updating process
    - Bugfix: [export] Export feed does not return the default image when is empty

=============================================================
Version 3.1.1
=============================================================

    - Feature: [import] Optimization of the order recovery system
    - Feature: [import] Setting up a cache for synchronizing catalogs ids
    - Feature: [action] Refactoring and optimization of actions on orders

=============================================================
Version 3.1.0
=============================================================

    - Feature: Disabling the Lengow tracker and changing the product ID
    - Feature: Registering marketplace data in a json file
    - Feature: Optimization of API calls between Magento and Lengow

app/code/community/Lengow/Connector/Helper/Sync.php
<?php
/**
 * Copyright 2017 Lengow SAS
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * @category    Lengow
 * @package     Lengow_Connector
 * @subpackage  Helper
 * @author      Team module <team-module@lengow.com>
 * @copyright   2017 Lengow SAS
 * @license     http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

/**
 * Helper sync
 */
class Lengow_Connector_Helper_Sync extends Mage_Core_Helper_Abstract
{
    /**
     * @var string cms type
     */
    const CMS_TYPE = 'magento';

    /* Sync actions */
    const SYNC_CATALOG = 'catalog';
    const SYNC_CMS_OPTION = 'cms_option';
    const SYNC_STATUS_ACCOUNT = 'status_account';
    const SYNC_MARKETPLACE = 'marketplace';
    const SYNC_ORDER = 'order';
    const SYNC_ACTION = 'action';
    const SYNC_PLUGIN_DATA = 'plugin';

    /**
     * @var string marketplace file name
     */
    const MARKETPLACE_FILE = 'marketplaces.json';

    /* Plugin link types */
    const LINK_TYPE_HELP_CENTER = 'help_center';
    const LINK_TYPE_CHANGELOG = 'changelog';
    const LINK_TYPE_UPDATE_GUIDE = 'update_guide';
    const LINK_TYPE_SUPPORT = 'support';

    /* Default plugin links */
    const LINK_HELP_CENTER = 'https://help.lengow.com/hc/en-us/articles/360011978332';
    const LINK_CHANGELOG = 'https://help.lengow.com/hc/en-us/articles/360011027659';
    const LINK_UPDATE_GUIDE = 'https://help.lengow.com/hc/en-us/articles/360011978332#12-update-the-plugin-version';
    const LINK_SUPPORT = 'https://help.lengow.com/hc/en-us/requests/new';

    /* Api iso codes */
    const API_ISO_CODE_EN = 'en';
    const API_ISO_CODE_FR = 'fr';
    const API_ISO_CODE_DE = 'de';

    /**
     * @var array cache time for catalog, account status, cms options and marketplace synchronisation
     */
    protected $_cacheTimes = array(
        self::SYNC_CATALOG => 21600,
        self::SYNC_CMS_OPTION => 86400,
        self::SYNC_STATUS_ACCOUNT => 86400,
        self::SYNC_MARKETPLACE => 43200,
        self::SYNC_PLUGIN_DATA => 86400,
    );

    /**
     * @var array valid sync actions
     */
    protected $_syncActions = array(
        self::SYNC_ORDER,
        self::SYNC_CMS_OPTION,
        self::SYNC_STATUS_ACCOUNT,
        self::SYNC_MARKETPLACE,
        self::SYNC_ACTION,
        self::SYNC_CATALOG,
        self::SYNC_PLUGIN_DATA,
    );

    /**
     * @var array iso code correspondence for plugin links
     */
    protected $_genericIsoCodes = array(
        self::API_ISO_CODE_EN => Lengow_Connector_Helper_Translation::ISO_CODE_EN,
        self::API_ISO_CODE_FR => Lengow_Connector_Helper_Translation::ISO_CODE_FR,
        self::API_ISO_CODE_DE => Lengow_Connector_Helper_Translation::ISO_CODE_DE,
    );

    /**
     * @var array default plugin links when the API is not available
     */
    protected $_defaultPluginLinks = array(
        self::LINK_TYPE_HELP_CENTER => self::LINK_HELP_CENTER,
        self::LINK_TYPE_CHANGELOG => self::LINK_CHANGELOG,
        self::LINK_TYPE_UPDATE_GUIDE => self::LINK_UPDATE_GUIDE,
        self::LINK_TYPE_SUPPORT => self::LINK_SUPPORT,
    );

    /**
     * @var Lengow_Connector_Helper_Config Lengow config helper instance
     */
    protected $_configHelper;

    /**
     * Construct
     */
    public function __construct()
    {
        $this->_configHelper = Mage::helper('lengow_connector/config');
    }

    /**
     * Is sync action
     *
     * @param string $action sync action
     *
     * @return boolean
     */
    public function isSyncAction($action)
    {
        return in_array($action, $this->_syncActions, true);
    }

    /**
     * Plugin is blocked or not
     *
     * @return boolean
     */
    public function pluginIsBlocked()
    {
        if ($this->_configHelper->isNewMerchant()) {
            return true;
        }
        $statusAccount = $this->getStatusAccount();
        return $statusAccount && ($statusAccount['type'] === 'free_trial' && $statusAccount['expired']);
    }


    /**
     * Get Sync Data (Inscription / Update)
     *
     * @return array
     */
    public function getSyncData()
    {
        /** @var Lengow_Connector_Helper_Data $helper */
        $helper = Mage::helper('lengow_connector');
        $data = array(
            'domain_name' => $_SERVER['SERVER_NAME'],
            'token' => $this->_configHelper->getToken(),
            'type' => self::CMS_TYPE,
            'version' => Mage::getVersion(),
            'plugin_version' => (string) Mage::getConfig()->getNode()->modules->Lengow_Connector->version,
            'email' => Mage::getStoreConfig('trans_email/ident_general/email'),
            'cron_url' => $helper->getCronUrl(),
            'toolbox_url' => $helper->getToolboxUrl(),
            'shops' => array(),
        );
        foreach (Mage::app()->getWebsites() as $website) {
            foreach ($website->getGroups() as $group) {
                $stores = $group->getStores();
                foreach ($stores as $store) {
                    $storeId = (int) $store->getId();
                    /** @var Lengow_Connector_Model_Export $export */
                    $export = Mage::getModel(
                        'lengow/export',
                        array(Lengow_Connector_Model_Export::PARAM_STORE_ID => $storeId)
                    );
                    $data['shops'][] = array(
                        'token' => $this->_configHelper->getToken($storeId),
                        'shop_name' => $store->getName(),
                        'domain_url' => $store->getBaseUrl(),
                        'feed_url' => $helper->getExportUrl($storeId),
                        'total_product_number' => $export->getTotalProduct(),
                        'exported_product_number' => $export->getTotalExportProduct(),
                        'enabled' => $this->_configHelper->storeIsActive($storeId),
                    );
                }
            }
        }
        return $data;
    }

    /**
     * Sync Lengow catalogs for order synchronisation
     *
     * @param boolean $force force cache update
     * @param boolean $logOutput see log or not
     *
     * @return boolean
     */
    public function syncCatalog($force = false, $logOutput = false)
    {
        $success = false;
        $cleanCache = false;
        if ($this->_configHelper->isNewMerchant()) {
            return $success;
        }
        if (!$force) {
            $updatedAt = $this->_configHelper->get(Lengow_Connector_Helper_Config::LAST_UPDATE_CATALOG);
            if ($updatedAt !== null && (time() - (int) $updatedAt) < $this->_cacheTimes[self::SYNC_CATALOG]) {
                return $success;
            }
        }
        $result = Mage::getModel('lengow/connector')->queryApi(
            Lengow_Connector_Model_Connector::GET,
            Lengow_Connector_Model_Connector::API_CMS,
            array(),
            '',
            $logOutput
        );
        if (isset($result->cms)) {
            $cmsToken = $this->_configHelper->getToken();
            foreach ($result->cms as $cms) {
                if ($cms->token === $cmsToken) {
                    foreach ($cms->shops as $cmsShop) {
                        $store = $this->_configHelper->getStoreByToken($cmsShop->token);
                        if ($store) {
                            $catalogIdsChange = $this->_configHelper->setCatalogIds(
                                $cmsShop->catalog_ids,
                                (int) $store->getId()
                            );
                            $activeStoreChange = $this->_configHelper->setActiveStore((int)$store->getId());
                            if (!$cleanCache && ($catalogIdsChange || $activeStoreChange)) {
                                $cleanCache = true;
                            }
                        }
                    }
                    $success = true;
                    break;
                }
            }
        }
        // clean config cache to valid configuration
        if ($cleanCache) {
            // save last update date for a specific settings (change synchronisation interval time)
            $this->_configHelper->set(Lengow_Connector_Helper_Config::LAST_UPDATE_SETTING, time());
            Mage::app()->getCacheInstance()->cleanType('config');
        }
        $this->_configHelper->set(Lengow_Connector_Helper_Config::LAST_UPDATE_CATALOG, time());
        return $success;
    }

    /**
     * Get options for all store
     *
     * @return array
     */
    public function getOptionData()
    {
        $data = array(
            'token' => $this->_configHelper->getToken(),
            'version' => Mage::getVersion(),
            'plugin_version' => (string) Mage::getConfig()->getNode()->modules->Lengow_Connector->version,
            'options' => $this->_configHelper->getAllValues(),
            'shops' => array(),
        );
        foreach (Mage::app()->getWebsites() as $website) {
            foreach ($website->getGroups() as $group) {
                $stores = $group->getStores();
                foreach ($stores as $store) {
                    $storeId = (int) $store->getId();
                    /** @var Lengow_Connector_Model_Export $export */
                    $export = Mage::getModel(
                        'lengow/export',
                        array(Lengow_Connector_Model_Export::PARAM_STORE_ID => $storeId)
                    );
                    $data['shops'][] = array(
                        'token' => $this->_configHelper->getToken($storeId),
                        'enabled' => $this->_configHelper->storeIsActive($storeId),
                        'total_product_number' => $export->getTotalProduct(),
                        'exported_product_number' => $export->getTotalExportProduct(),
                        'options' => $this->_configHelper->getAllValues($storeId),
                    );
                }
            }
        }
        return $data;
    }

    /**
     * Set CMS options
     *
     * @param boolean $force force cache update
     * @param boolean $logOutput see log or not
     *
     * @return boolean
     */
    public function setCmsOption($force = false, $logOutput = false)
    {
        if ($this->_configHelper->isNewMerchant() || $this->_configHelper->debugModeIsActive()) {
            return false;
        }
        if (!$force) {
            $updatedAt = $this->_configHelper->get(Lengow_Connector_Helper_Config::LAST_UPDATE_OPTION_CMS);
            if ($updatedAt !== null && (time() - (int) $updatedAt) < $this->_cacheTimes[self::SYNC_CMS_OPTION]) {
                return false;
            }
        }
        $options = Mage::helper('core')->jsonEncode($this->getOptionData());
        Mage::getModel('lengow/connector')->queryApi(
            Lengow_Connector_Model_Connector::PUT,
            Lengow_Connector_Model_Connector::API_CMS,
            array(),
            $options,
            $logOutput
        );
        $this->_configHelper->set(Lengow_Connector_Helper_Config::LAST_UPDATE_OPTION_CMS, time());
        return true;
    }

    /**
     * Get Status Account
     *
     * @param boolean $force force cache update
     * @param boolean $logOutput see log or not
     *
     * @return array|false
     */
    public function getStatusAccount($force = false, $logOutput = false)
    {
        if ($this->_configHelper->isNewMerchant()) {
            return false;
        }
        if (!$force) {
            $updatedAt = $this->_configHelper->get(Lengow_Connector_Helper_Config::LAST_UPDATE_ACCOUNT_STATUS_DATA);
            if ($updatedAt !== null && (time() - (int) $updatedAt) < $this->_cacheTimes[self::SYNC_STATUS_ACCOUNT]) {
                return json_decode(
                    $this->_configHelper->get(Lengow_Connector_Helper_Config::ACCOUNT_STATUS_DATA),
                    true
                );
            }
        }
        $result = Mage::getModel('lengow/connector')->queryApi(
            Lengow_Connector_Model_Connector::GET,
            Lengow_Connector_Model_Connector::API_PLAN,
            array(),
            '',
            $logOutput
        );
        if (isset($result->isFreeTrial)) {
            $status = array(
                'type' => $result->isFreeTrial ? 'free_trial' : '',
                'day' => (int) $result->leftDaysBeforeExpired < 0 ? 0 : (int) $result->leftDaysBeforeExpired,
                'expired' => (bool) $result->isExpired,
                'legacy' => $result->accountVersion === 'v2',
            );
            $this->_configHelper->set(
                Lengow_Connector_Helper_Config::ACCOUNT_STATUS_DATA,
                Mage::helper('core')->jsonEncode($status)
            );
            $this->_configHelper->set(Lengow_Connector_Helper_Config::LAST_UPDATE_ACCOUNT_STATUS_DATA, time());
            return $status;
        }
        if ($this->_configHelper->get(Lengow_Connector_Helper_Config::ACCOUNT_STATUS_DATA)) {
            return json_decode($this->_configHelper->get(Lengow_Connector_Helper_Config::ACCOUNT_STATUS_DATA), true);
        }
        return false;
    }

    /**
     * Get marketplace data
     *
     * @param boolean $force force cache update
     * @param boolean $logOutput see log or not
     *
     * @return array|false
     */
    public function getMarketplaces($force = false, $logOutput = false)
    {
        $folderPath = Mage::getModuleDir('etc', 'Lengow_Connector');
        $filePath = $folderPath . DIRECTORY_SEPARATOR . self::MARKETPLACE_FILE;
        if (!$force) {
            $updatedAt = $this->_configHelper->get(Lengow_Connector_Helper_Config::LAST_UPDATE_MARKETPLACE);
            if ($updatedAt !== null
                && (time() - (int) $updatedAt) < $this->_cacheTimes[self::SYNC_MARKETPLACE]
                && file_exists($filePath)
            ) {
                // recovering data with the marketplaces.json file
                $marketplacesData = file_get_contents($filePath);
                if ($marketplacesData) {
                    return json_decode($marketplacesData);
                }
            }
        }
        // recovering data with the API
        $result = Mage::getModel('lengow/connector')->queryApi(
            Lengow_Connector_Model_Connector::GET,
            Lengow_Connector_Model_Connector::API_MARKETPLACE,
            array(),
            '',
            $logOutput
        );
        if ($result && is_object($result) && !isset($result->error)) {
            // updated marketplaces.json file
            try {
                $file = new Varien_Io_File();
                $file->cd($folderPath);
                $file->streamOpen(self::MARKETPLACE_FILE, 'w+');
                $file->streamlock();
                $file->streamWrite(json_encode($result));
                $file->streamUnlock();
                $file->streamClose();
                $this->_configHelper->set(Lengow_Connector_Helper_Config::LAST_UPDATE_MARKETPLACE, time());
            } catch (Exception $e) {
                $helper = Mage::helper('lengow_connector/data');
                $helper->log(
                    Lengow_Connector_Helper_Data::CODE_IMPORT,
                    $helper->setLogMessage(
                        'log.import.marketplace_update_failed',
                        array('error_message' => $e->getMessage())
                    ),
                    $logOutput
                );
            }
            return $result;
        }
        // if the API does not respond, use marketplaces.json if it exists
        if (file_exists($filePath)) {
            $marketplacesData = file_get_contents($filePath);
            if ($marketplacesData) {
                // don't add true, decoded data are used as object
                return json_decode($marketplacesData);
            }
        }
        return false;
    }

    /**
     * Get Lengow plugin data (last version and download link)
     *
     * @param boolean $force force cache update
     * @param boolean $logOutput see log or not
     *
     * @return array|false
     */
    public function getPluginData($force = false, $logOutput = false)
    {
        if (!$force) {
            $updatedAt = $this->_configHelper->get(Lengow_Connector_Helper_Config::LAST_UPDATE_PLUGIN_DATA);
            if ($updatedAt !== null && (time() - (int) $updatedAt) < $this->_cacheTimes[self::SYNC_PLUGIN_DATA]) {
                return json_decode($this->_configHelper->get(Lengow_Connector_Helper_Config::PLUGIN_DATA), true);
            }
        }
        $plugins = Mage::getModel('lengow/connector')->queryApi(
            Lengow_Connector_Model_Connector::GET,
            Lengow_Connector_Model_Connector::API_PLUGIN,
            array(),
            '',
            $logOutput
        );
        if ($plugins) {
            $pluginData = false;
            foreach ($plugins as $plugin) {
                if ($plugin->type === self::CMS_TYPE) {
                    $cmsMinVersion = '';
                    $cmsMaxVersion = '';
                    $pluginLinks = array();
                    $currentVersion = $plugin->version;
                    if (!empty($plugin->versions)) {
                        foreach ($plugin->versions as $version) {
                            if ($version->version === $currentVersion) {
                                $cmsMinVersion = $version->cms_min_version;
                                $cmsMaxVersion = $version->cms_max_version;
                                break;
                            }
                        }
                    }
                    if (!empty($plugin->links)) {
                        foreach ($plugin->links as $link) {
                            if (array_key_exists($link->language->iso_a2, $this->_genericIsoCodes)) {
                                $genericIsoCode = $this->_genericIsoCodes[$link->language->iso_a2];
                                $pluginLinks[$genericIsoCode][$link->link_type] = $link->link;
                            }
                        }
                    }
                    $pluginData = array(
                        'version' => $currentVersion,
                        'download_link' => $plugin->archive,
                        'cms_min_version' => $cmsMinVersion,
                        'cms_max_version' => $cmsMaxVersion,
                        'links' => $pluginLinks,
                        'extensions' => $plugin->extensions,
                    );
                    break;
                }
            }
            if ($pluginData) {
                $this->_configHelper->set(
                    Lengow_Connector_Helper_Config::PLUGIN_DATA,
                    Mage::helper('core')->jsonEncode($pluginData)
                );
                $this->_configHelper->set(Lengow_Connector_Helper_Config::LAST_UPDATE_PLUGIN_DATA, time());
                return $pluginData;
            }
        } elseif ($this->_configHelper->get(Lengow_Connector_Helper_Config::PLUGIN_DATA)) {
            return json_decode($this->_configHelper->get(Lengow_Connector_Helper_Config::PLUGIN_DATA), true);
        }
        return false;
    }

    /**
     * Get an array of plugin links for a specific iso code
     *
     * @param string|null $isoCode
     *
     * @return array
     */
    public function getPluginLinks($isoCode = null)
    {
        $pluginData = $this->getPluginData();
        if (!$pluginData) {
            return $this->_defaultPluginLinks;
        }
        // check if the links are available in the locale
        $isoCode = $isoCode ?: Lengow_Connector_Helper_Translation::DEFAULT_ISO_CODE;
        $localeLinks = isset($pluginData['links'][$isoCode]) ? $pluginData['links'][$isoCode] : false;
        $defaultLocaleLinks = isset($pluginData['links'][Lengow_Connector_Helper_Translation::DEFAULT_ISO_CODE])
            ? $pluginData['links'][Lengow_Connector_Helper_Translation::DEFAULT_ISO_CODE]
            : false;
        // for each type of link, we check if the link is translated
        $pluginLinks = array();
        foreach ($this->_defaultPluginLinks as $linkType => $defaultLink) {
            if ($localeLinks && isset($localeLinks[$linkType])) {
                $link = $localeLinks[$linkType];
            } elseif ($defaultLocaleLinks && isset($defaultLocaleLinks[$linkType])) {
                $link = $defaultLocaleLinks[$linkType];
            } else {
                $link = $defaultLink;
            }
            $pluginLinks[$linkType] = $link;
        }
        return $pluginLinks;
    }
}
